File: app/src/main/java/com/pluu/support/naver/NaverDetailApi.java
package com.pluu.support.naver;

import android.content.Context;
import android.text.TextUtils;

import com.annimon.stream.Stream;
import com.pluu.support.impl.AbstractDetailApi;
import com.pluu.webtoon.item.Detail;
import com.pluu.webtoon.item.DetailView;
import com.pluu.webtoon.item.ERROR_TYPE;
import com.pluu.webtoon.item.Episode;
import com.pluu.webtoon.item.ShareItem;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.List;

/**
 * 네이버 웹툰 상세 API
 * Created by PLUUSYSTEM-NEW on 2015-10-29.
 */
public class NaverDetailApi extends AbstractDetailApi {

    private final String SHARE_URL = "http://m.comic.naver.com/webtoon/detail.nhn?titleId=%s&no=%s";
    private final List<String> SKIP_DETAIL = new ArrayList<String>() {{
        add("http://static.naver.com/m/comic/im/txt_ads.png");
        add("http://static.naver.com/m/comic/im/toon_app_pop.png");
    }};

    private String webToonId, episodeId;

    public NaverDetailApi(Context context) {
        super(context);
    }

    @Override
    public Detail parseDetail(Episode episode) {
        this.webToonId = episode.getToonId();
        this.episodeId = episode.getEpisodeId();

        Detail ret = new Detail();
        ret.webtoonId = webToonId;
        ret.episodeId = episodeId;

        String response;
        try {
            response = requestApi();
        } catch (Exception e) {
            e.printStackTrace();
            ret.list = new ArrayList<>();
            return ret;
        }

        Document doc = Jsoup.parse(response);
        ret.title = doc.select("div[class=chh] span, h1[class=tit]").first().text();

        Elements tempElements;

        tempElements = doc.select("div[class=viewer cuttoon]");
        if (tempElements != null && !tempElements.isEmpty()) {
            // 컷툰
            parseCutToon(ret, doc);
            return ret;
        }

        tempElements = doc.select(".oz-loader");
        if (tempElements != null && !tempElements.isEmpty()) {
            // osLoader
            ret.errorType = ERROR_TYPE.NOT_SUPPORT;
            return ret;
        }

        // 일반 웹툰
        parseNormal(ret, doc);

        return ret;
    }

    private void parseNormal(Detail ret, Document doc) {
        ret.list = parseDetailNormalType(doc);

        // 이전, 다음화
        Elements navi = doc.select(".navi_area");
        Elements temp = navi.select("[data-type=next]");
        if (temp != null && !temp.isEmpty()) {
            ret.nextLink = temp.attr("data-no");
        }
        temp = navi.select("[data-type=prev]");
        if (temp != null && !temp.isEmpty()) {
            ret.prevLink = temp.attr("data-no");
        }
    }

    private void parseCutToon(Detail ret, Document doc) {
        ret.list = parseDetailCutToonType(doc);

        // 이전, 다음화
        Elements navi = doc.select(".paging_wrap");
        Elements temp = navi.select("[data-type=next]");
        if (temp != null && !temp.isEmpty()) {
            ret.nextLink = temp.attr("data-no");
        }
        temp = navi.select("[data-type=prev]");
        if (temp != null && !temp.isEmpty()) {
            ret.prevLink = temp.attr("data-no");
        }
    }

    private List<DetailView> parseDetailCutToonType(Document doc) {
        return Stream.of(doc.select(".swiper-slide img.swiper-lazy"))
            .map(element -> DetailView.createImage(element.attr("data-src")))
            .toList();
    }

    private List<DetailView> parseDetailNormalType(Document doc) {
        List<DetailView> list = new ArrayList<>();
        Elements links = doc.select("#toonLayer li img");
        String path;
        for (Element item : links) {
            path = item.attr("data-original");
            if (TextUtils.isEmpty(path)) {
                path = item.attr("src");
            }
            if (!TextUtils.isEmpty(path)
                && !SKIP_DETAIL.contains(path)) {
                list.add(DetailView.createImage(path));
            }
        }
        return list;
    }

    @Override
    public ShareItem getDetailShare(Episode episode, Detail detail) {
        String url = String.format(SHARE_URL, detail.webtoonId, detail.episodeId);
        ShareItem item = new ShareItem();
        item.title = episode.getTitle() + " / " + detail.title;
        item.url = url;
        return item;
    }

    @Override
    public String getMethod() {
        return GET;
    }

    @Override
    public String getUrl() {
        return String.format(SHARE_URL, webToonId, episodeId);
    }
}

File: app/src/main/java/com/pluu/support/naver/NaverDetailApi.kt
package com.pluu.support.naver

import com.pluu.support.impl.AbstractDetailApi
import com.pluu.support.impl.REQUEST_METHOD
import com.pluu.webtoon.di.NetworkUseCase
import com.pluu.webtoon.item.*
import org.jsoup.Jsoup
import org.jsoup.nodes.Document

/**
 * 네이버 웹툰 상세 API
 * Created by pluu on 2017-04-20.
 */
class NaverDetailApi(
    networkUseCase: NetworkUseCase
) : AbstractDetailApi(networkUseCase) {

    private val SKIP_DETAIL = arrayOf(
        "http://static.naver.com/m/comic/im/txt_ads.png",
        "http://static.naver.com/m/comic/im/toon_app_pop.png"
    )

    private lateinit var webToonId: String
    private lateinit var episodeId: String

    override fun parseDetail(episode: Episode): Detail {
        this.webToonId = episode.toonId
        this.episodeId = episode.episodeId

        val ret = Detail().apply {
            webtoonId = episode.toonId
            episodeId = episode.episodeId
        }

        val doc = try {
            Jsoup.parse(requestApi())
        } catch (e: Exception) {
            e.printStackTrace()
            ret.list = emptyList()
            return ret
        }

        ret.title = doc.select("div[class=chh] span, h1[class=tit]").first()?.text()

        when {
            doc.select("#ct")?.isNotEmpty() == true -> {
                // 컷툰
                parseCutToon(ret, doc)
            }
            doc.select(".oz-loader")?.isNotEmpty() == true -> {
                // osLoader
                ret.errorType = ERROR_TYPE.NOT_SUPPORT
            }
            // 일반 웹툰
            else -> parseNormal(ret, doc)
        }
        return ret
    }

    private fun parseNormal(ret: Detail, doc: Document) {
        ret.list = parseDetailNormalType(doc)

        // 이전, 다음화
        doc.select(".paging_wrap").apply {
            select("[data-type=next]").takeIf { it.isNotEmpty() }.let {
                ret.nextLink = (episodeId.toInt() + 1).toString()
            }
            select("[data-type=prev]").takeIf { it.isNotEmpty() }.let {
                ret.prevLink = (episodeId.toInt() - 1).toString()
            }
        }
    }

    private fun parseCutToon(ret: Detail, doc: Document) {
        ret.list = parseDetailCutToonType(doc)

        // 이전, 다음화
//        doc.select(".paging_wrap")?.apply {
//            select("[data-type=next]")?.attr("data-no")?.apply {
//                ret.nextLink = if (isNotEmpty()) this else null
//            }
//            select("[data-type=prev]")?.attr("data-no")?.apply {
//                ret.prevLink = if (isNotEmpty()) this else null
//            }
//        }
    }

    private fun parseDetailCutToonType(doc: Document): List<DetailView> {
        return doc.select("#ct ul li img")
            .map { it -> it.attr("data-src") }
            .map { url -> DetailView(url) }
    }

    private fun parseDetailNormalType(doc: Document) =
        doc.select("#toonLayer li img")
            .map { it ->
                it.attr("data-original").takeIf {
                    it.isNotEmpty()
                } ?: it.attr("src")
            }
            .filter { it.isNotEmpty() && !SKIP_DETAIL.contains(it) }
            .map { DetailView(it) }

    private fun getShareUrl(webtoonId: String, episodeId: String) =
        "http://m.comic.naver.com/webtoon/detail.nhn?titleId=$webtoonId&no=$episodeId"

    override fun getDetailShare(episode: Episode, detail: Detail) = ShareItem(
        title = "${episode.title} / ${detail.title}",
        url = getShareUrl(detail.webtoonId.orEmpty(), detail.episodeId.orEmpty())
    )

    override val method: REQUEST_METHOD = REQUEST_METHOD.GET

    override val url: String
        get() = getShareUrl(webToonId, episodeId)
}
